crates/sui-indexer/src/config.rs

```diff
@@ -208,8 +214,93 @@ pub enum Command {
+/// not have an override specified. Instantiated only if `PruningOptions` is provided on indexer
+/// start.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RetentionPolicies {
+    /// Default retention policy for all tables.
+    pub epochs_to_keep: u64,
+    /// In Rust, a mapping of all `PrunableTable` variants to their respective retention policies.
+    /// From a TOML file, a user only needs to explicitly list tables that should override the
+    /// default retention.
+    #[serde(default, rename = "overrides")]
+    pub policies: HashMap<PrunableTable, u64>,
+}
+
+impl PruningOptions {
+    /// Load default retention policy and overrides from file.
+    pub fn load_from_file(&self) -> Option<RetentionPolicies> {
+        let Some(config_path) = self.pruning_config_path.as_ref() else {
+            return None;
+        };
+
+        let contents = std::fs::read_to_string(&config_path)
+            .expect("Failed to read default retention policy and overrides from file");
+        let retention_with_overrides = toml::de::from_str::<RetentionPolicies>(&contents)
+            .expect("Failed to parse into RetentionPolicies struct");
+
+        let default_retention = retention_with_overrides.epochs_to_keep;
+
+        assert!(
+            default_retention > 0,
+            "Default retention must be greater than 0"
+        );
+        assert!(
+            retention_with_overrides
+                .policies
+                .values()
+                .all(|&policy| policy > 0),
+            "All retention overrides must be greater than 0"
+        );
+
+        Some(retention_with_overrides)
+    }
+}
+
+impl RetentionPolicies {
+    /// Create a new `RetentionPolicies` with the specified default retention and overrides. Call
+    /// `finalize()` on the instance to update the `policies` field with the default retention
+    /// policy for all tables that do not have an override specified.
+    pub fn new(epochs_to_keep: u64, overrides: HashMap<PrunableTable, u64>) -> Self {
+        Self {
+            epochs_to_keep,
+            policies: overrides,
+        }
+    }
+
+    /// Create a new `RetentionPolicies` with only the default retention specified and the default
+    /// override for `objects_history`. Call `finalize()` on the instance to update the `policies`
+    /// field with the default retention policy for all tables that do not have an override
+    /// specified.
+    pub fn new_with_default_retention_only_for_testing(epochs_to_keep: u64) -> Self {
+        let mut overrides = HashMap::new();
+        overrides.insert(
+            PrunableTable::ObjectsHistory,
+            OBJECTS_HISTORY_EPOCHS_TO_KEEP,
+        );
+
+        Self::new(epochs_to_keep, overrides)
+    }
+
+    /// Updates the `policies` field with the default retention policy for all tables that do not
+    /// have an override specified.
+    pub fn finalize(mut self) -> Self {
```

Comment:
I find it a bit odd to have both a `new` and a `finalize` method when no changes are made between their calls. I understand the purpose might be to reuse the override config map as a complete set, combining both default and override pruning configurations. But what do you think of simplifying this by keeping the struct as just the config, and having a function that takes the config and returns a hashmap with the complete retention policies like this?
```rust
pub struct RetentionConfig {
    /// Default retention policy for all tables.
    pub epochs_to_keep: u64,
    /// A user only needs to explicitly list tables that should override the
    /// default retention.
    #[serde(default)]
    pub overrides: HashMap<PrunableTable, u64>,
}

pub fn retention_policies_from_config(config: RetentionConfig
) -> HashMap<PrunableTable, u64> {
    let RetentionConfig {
        epochs_to_keep, overrides
    }
    let res = overrides;
    for table in PrunableTable::iter() {
        res.entry(table).or_insert(self.epochs_to_keep);
    }
    res
}
```